<a href="https://colab.research.google.com/github/R0onald/Analise_fraude-Machine-learning/blob/main/Detec%C3%A7%C3%A3oDeFraudeComML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

****Detecção de fraude com aprendizado de máquina****

- vamos começar importando todas as bibliotecas necessárias para detecção de fraude com aprendizado de máquina

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

**Modelo de detecção de fraude de pagamento**

O conjunto de dados que estou usando já está muito bem estruturado, sem valores ausentes e não encontro nenhum escopo de limpeza de dados nele. Portanto, sem perder tempo, irei mergulhar na construção de nosso modelo de aprendizado de máquina.

In [ ]:
from google.colab import files
uploaded = files.upload()
df = pd.read_csv('payment_fraud.txt')
df.head()

Saving payment_fraud.txt to payment_fraud (1).txt


accountAgeDays  numItems  localTime paymentMethod  paymentMethodAgeDays  \
0              29         1   4.745402        paypal             28.204861   
1             725         1   4.742303   storecredit              0.000000   
2             845         1   4.921318    creditcard              0.000000   
3             503         1   4.886641    creditcard              0.000000   
4            2000         1   5.040929    creditcard              0.000000   

   label  
0      0  
1      0  
2      0  
3      0  
4      0

- Agora, dividirei os dados em conjuntos de treinamento e teste:

In [ ]:
# Split dataset up into train and test sets
X_train, X_test, y_train, y_test = train_test_split(
   df.drop('label', axis=1), df['label'], test_size=0.33, random_state=17)

Como se trata de um problema de classificação binária, utilizarei um algoritmo de Regressão Logística

- Eu vou precisar fazer uma pequena análise exploratória para resolver um bug no modelo

In [ ]:

non_numeric_cols = X_train.select_dtypes(include='object').columns
print(non_numeric_cols)

Index(['paymentMethod'], dtype='object')


In [ ]:
print(X_train.dtypes)
print(y_train.dtypes)

accountAgeDays            int64
numItems                  int64
localTime               float64
paymentMethod            object
paymentMethodAgeDays    float64
dtype: object
int64


In [ ]:

for col in non_numeric_cols:
    print(X_train[col].value_counts())

creditcard     18666
paypal          6323
storecredit     1289
Name: paymentMethod, dtype: int64


In [ ]:
X_train = X_train.drop(non_numeric_cols, axis=1)
# or
X_train = pd.get_dummies(X_train, drop_first=True)

In [ ]:
clf = LogisticRegression().fit(X_train, y_train)

# Make predictions on test set
y_pred = clf.predict(X_test)
from sklearn.metrics import accuracy_score
print(accuracy_score(y_pred, y_test))

1.0


Nosso modelo de detecção de fraude obteve uma precisão de 100% usando o algoritmo de regressão logística.

**Avaliando o modelo de detecção de fraude**

- Usarei o algoritmo da matriz de confusão para avaliar o desempenho do nosso modelo. Podemos usar o algoritmo da matriz de confusão apenas com um código de uma linha:

In [ ]:
# Compare test set predictions with ground truth labels
print(confusion_matrix(y_test, y_pred))

[[12753     0]
 [    0   190]]


Portanto, de todas as transações no conjunto de dados, 190 transações são corretamente reconhecidas como fraude e 12.753 transações são reconhecidas como transações não fraudulentas.